### ISSUE TITLE
Suggestion: Warnings about unintuitive "broken" behaviour when combining default interface methods, generics, and value types

### ISSUE THREAD

**Issue by SamPruden** [dotnet/roslyn-analyzers#5587]
### Describe the problem you are trying to solve
Combining default interface methods, generics, and value types results in an unintuitive boxed copy that interferes with expected reference semantics.

```C#
var foo = new Foo();
GenericallyIncrement(ref foo);
Console.WriteLine(foo.Bar); // Not actually incremented!!!

void GenericallyIncrement<T>(ref T foo) where T : IFoo => foo.IncrementThroughInterface();

interface IFoo
{
    int Bar { get; set; }
    void IncrementThroughInterface() => this.Bar++;
}

struct Foo : IFoo
{
    public int Bar { get; set; }
}
```
[SharpLab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgNwEMwACAMwHtLSBeUqAUwHdSAxagCgEoBuPAOKMmYAJYBjIgBspATwCSUcWEYBbYcE4ryFanzypMATk5VKAOgBCJPqQD0d0gDlKwUqKUr1UYIzh4DABZSIREJaTlFZTUNAB4AFQA+LUYdeN1KblJmAAtGFVJ0kFJ5Dko8Ukq6RIzzKK8NeJywSggAcxzFXzByInFGHn58XA9u3v6SsrwAbwqq0dJrMmnSNsZgXlIAZ3XNgF85ytRg+pifJpb2zp988YGs2hrgHNEtqxJERCGD4a3gSHEbjKpGKpWoM0OpFQAGZ3D5FiRSCs1httrtSD89kA==)

[CSharpLang Discussion](https://github.com/dotnet/csharplang/discussions/4885)

This is unexpected when intuitively combining the features, and quite hard to debug. In my case, when I first discovered this I got quite far into an architectural design that relied on this working before I noticed the problem, and had to scrap a day or so's work to refactor around it.

I actually strongly dislike that this issue exists and wish it would go away, but based on the CSharpLang discussion it looks like it's here to stay. I believe this issue actually breaks the "seamlessly add new methods to interfaces" intent behind DIM, because of the fairly complicated ways in which this issue can crop up and catch you unawares when doing so, although admittedly most code bases probably don't combine interfaces with structs that much. I worry that it's in the sweet spot of being rare enough that people may not know to look out for it, and unintuitive edge case bugs could ship because of it.

A good analyzer or set of analyzers should robustly catch this issue when it occurs. If we imagine that the interface is declared in a library, and implemented on a struct in user code, then both the library author and consumer should be warned about this where it may be dangerous. The generic method that consumes the interface could be implemented in either the library or user code.

### Describe suggestions on how to achieve the rule

This suggestion is a little vague because I think there needs to be some discussion about how best to use analyzers to warn against this. In the above code sample, there are a few places where we could warn.

#### Warn on generic method call
The `GenericallyIncrement(ref foo)` call could warn, because it's calling a generic method that's constrained to an interface, and `T` has members of that interface implemented by DIM. However, this isn't always a problem. The issue only occurs if the generic method implementation uses the DIM methods - it could be that that interface has another non-DIM method that is used instead without issue.

Even if a DIM method is used, the issue varies.  I believe that it will always cause an unexpected boxing, which should probably be warned about for performance reasons. However, the unexpected behaviour will only happen if the called DIM method attempts to mutate itself.

#### Warn on DIM method call
`foo.IncrementThroughInterface()` could warn, because `T` might be a struct, and if it is a struct, it might not have overridden the DIM. One possible code-fix here would be "Add a class constraint." However, if `T` does override the DIM then there isn't an issue here and this should be allowed without warning. Retroactively adding `class` constraints to APIs to accommodate DIMs would interfere with the "seamlessly add methods to interfaces" goal of DIMs.

Again, the issue here varies between just unexpected slow boxing and actual unexpected behaviour, depending on whether the DIM method mutates itself.

#### Warn on struct declaration
The struct declaration could warn that "Structs should always override DIM." This makes sense, although it defeats the purpose of DIM in this case. The idea is that one should be able to add a new method to an interface without breaking things that implement it, however adding a new DIM method to an interface would now "break" any structs that implement it by adding a warning to them.

---

Because of the library issue discussed above, it may make sense to warn in all of these places, although none of them are ideal. The warnings may catch a lot of false positives and appear spammy.

Each of the above rules has a potential variation where it actually checks the DIM body to see if it mutates itself. This would be a bit more targeted, but potentially complicated and slow to implement in an analyzer? There should probably be a performance warning about the unintuitive boxing even if it doesn't mutate itself.

**Comment by SamPruden** [comment]
Tagging in @CyrusNajmabadi because he agreed in the CSharpLang discussion that an analyzer might be a good idea, and might have some good thoughts on what the user experience should be.
